### DISCUSSION TITLE
Live Demo / Website

### DISCUSSION THREAD

**Discussion by jogibear9988** [saasquatch/raisins#13] [General]
Hey, I've done a similar project https://node-projects.github.io/web-component-designer-demo/index.html, also using and supporting webcomponents.

Is there a live demo of this anywhere? I'd like to see how it looks like.

**Comment by loganvolkers** [comment]
Not yet! There's some project hygiene underway, and that will be followed by some publishing of storybooks, website and docs.

**Reply by jogibear9988** [reply]
I'm very interested in what it can do and how it is comparable to my solution. How long du you think it will take before a demo is ready?

**Reply by loganvolkers** [reply]
The best analogy is to look at GrapesJS. This started as an upgrade to Grapes, turned into a fork, then a ground up rewrite using Jotai as the atomic state solution. 

This project also highlighted the need for and birthed [bunshi](https://bunshi.org).

Compared to your library, raisins is targeted at non-technical users, and high levels of security (the canvas iframe is sandboxed) to allow for safer use of 3rd party scripts. 

The single most important feature that raisins provides over grapes is the iframe sandboxing.

**Reply by jogibear9988** [reply]
Doesn't GrapeJS not also use a iframe?

Mine could also be used directly as a webcomponent without much setup, but the usage I think of is more to include it in your app.
(for example a simple usage: https://github.com/node-projects/web-component-designer-simple-demo)

I also did not use a iframe, cause then every component I want to design with needs to be loaded also in the iframe, and for example I a use case in our company, this would be much more load time.
But I also thought if I want to support both, iframe and shadowDom usage, maybe I will work on this when I've time.

Also in a early version in my project, I highlighted the elements directly in the designer (via adding css classes, or properties), but this created problems and so I switched to an extra overlay layer component.

**Reply by loganvolkers** [reply]
There's a big difference between an iframe and a sandboxed iframe.

With an iframe any html that you load could load JavaScript that causes a "cross site scripting" attack (even though it's not really cross-site, it's just a scripting attack).

This was highlighted during penetration testing of GrapesJS. That's why raisins got started. 

Sandboxing prevents these kinds of attacks, but requires rewriting the canvas, which is the hardest thing to work on.

**Reply by jogibear9988** [reply]
Oh, never knew this exists... (https://www.w3schools.com/tags/att_iframe_sandbox.asp)

I now tried in my designer to switch from ShadowRoot to iframe, basic editing works, but it will need more work. I'll look to get this fixed, real iframe means, I also could support real documents as children. 
The use of sandbox attribute hopefully would also work on mine, will try

**Reply by loganvolkers** [reply]
You'll likely need a [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) solution for cross-origin communication for things like events. The best is a little-known library called [Penpal](https://github.com/Aaronius/penpal).

**Reply by jogibear9988** [reply]
Why would I need events from the design surface? I do every clicking and overlay from outside.

I now started to work on support: https://github.com/node-projects/web-component-designer/pull/424/files#diff-0831c97952cec3ebcdce048e39d951e7b8a42b8a12902f35b449c75994cac51dR345 (I need the "allow-same-origin" in the sandbox attribute, is it still secure?)

**Reply by jogibear9988** [reply]
If it is interesting for you, I've now added iframe support:

<img width="132" alt="image" src="https://github.com/saasquatch/raisins/assets/364896/6d276636-2c45-454d-ac76-c79de8297239">

iframe is sandboxed like this:

<img width="342" alt="image" src="https://github.com/saasquatch/raisins/assets/364896/4f2f624c-516f-406e-80e9-a5d60a89bc45">

there are still some features not corretly working, cause I need to calculate the offsets and mouse positions relative to the frame...
but this a WIP

**Reply by jogibear9988** [reply]
And also on many places where I called APIs of the window, I need to use the API of the iframe window... 
I've not yet fixed all

**Reply by loganvolkers** [reply]
Try putting a XSS attack into your iframe, and see if it works, for example:

```html
<script>
window.parent.innerHTML = ""
</script>
```

You can see that JsFiddle works the right way. The iframe is isolated from overriding the secure parent context: https://jsfiddle.net/nrs3ju2z/

It throws the appropriate error in the browsers.

![image](https://github.com/saasquatch/raisins/assets/1157086/79a8cf16-237d-4dd4-8fe4-75c9f23a5200)

Since you are using `allow-same-origin`, this breaks the sandbox. Now any script in the frame can access and make any changes to the parent frame.

**Reply by jogibear9988** [reply]
No, this will not work in mine, cause I did not set "allow-scripts"

<img width="598" alt="image" src="https://github.com/saasquatch/raisins/assets/364896/2e564d65-2506-49dc-a96e-fabbbd706393">

But if I'd like to allow them, I need to do a communication to the iframe as you said. But this then also means, I need to inject a script into the iframe.
Maybe I look to support this also, will try how much work it is.

**Reply by jogibear9988** [reply]
I also tried if I could prevent access via this script:

     delete window.parent;
     delete window.frameElement;
     delete window.top;

but window.top seems not to be overridable
